File: modules/kafka/src/main/scala/com/dimafeng/testcontainers/KafkaContainer.scala
package com.dimafeng.testcontainers

import org.testcontainers.kafka.{KafkaContainer => JavaKafkaContainer}
import org.testcontainers.utility.DockerImageName

case class KafkaContainer(dockerImageName: DockerImageName = DockerImageName.parse(KafkaContainer.defaultDockerImageName)
                    ) extends SingleContainer[JavaKafkaContainer] {

  override val container: JavaKafkaContainer = new JavaKafkaContainer(dockerImageName)

  def bootstrapServers: String = container.getBootstrapServers
}

object KafkaContainer {

  val defaultDockerImageName = "apache/kafka"

  case class Def(dockerImageName: DockerImageName = DockerImageName.parse(defaultDockerImageName)
                ) extends ContainerDef {

    override type Container = KafkaContainer

    override def createContainer(): KafkaContainer = {
      new KafkaContainer(dockerImageName)
    }
  }
}

File: modules/kafka/src/main/scala/com/dimafeng/testcontainers/KafkaContainer.java
package com.dimafeng.testcontainers;

class KafkaContainer(confluentPlatformVersion: Option[String] = None,
                     externalZookeeper: Option[String] = None) extends SingleContainer[OTCGenericContainer[_]] {

  type OTCContainer = OTCGenericContainer[T] forSome {type T <: OTCKafkaContainer}

  val kafkaContainer: OTCKafkaContainer = {
    if (confluentPlatformVersion.isEmpty) {
      new OTCKafkaContainer()
    } else {
      new OTCKafkaContainer(confluentPlatformVersion.get)
    }
  }

  if (externalZookeeper.isEmpty) {
    kafkaContainer.withEmbeddedZookeeper()
  } else {
    kafkaContainer.withExternalZookeeper(externalZookeeper.get)
  }

  override val container: OTCKafkaContainer = kafkaContainer

}
